pkg/apis/build/v1alpha1/image_validation.go

```diff
@@ -26,46 +27,61 @@ func init() {
 	defaultCacheSize = resource.MustParse("2G")
 }
 
-func (s *Image) SetDefaults(ctx context.Context) {
-	if s.Spec.ServiceAccount == "" {
-		s.Spec.ServiceAccount = "default"
+func (i *Image) SetDefaults(ctx context.Context) {
+	if i.Spec.ServiceAccount == "" {
+		i.Spec.ServiceAccount = "default"
 	}
 
-	if s.Spec.ImageTaggingStrategy == "" {
-		s.Spec.ImageTaggingStrategy = BuildNumber
+	if i.Spec.ImageTaggingStrategy == "" {
+		i.Spec.ImageTaggingStrategy = BuildNumber
 	}
 
-	if s.Spec.FailedBuildHistoryLimit == nil {
-		s.Spec.FailedBuildHistoryLimit = &defaultFailedBuildHistoryLimit
+	if i.Spec.FailedBuildHistoryLimit == nil {
+		i.Spec.FailedBuildHistoryLimit = &defaultFailedBuildHistoryLimit
 	}
 
-	if s.Spec.SuccessBuildHistoryLimit == nil {
-		s.Spec.SuccessBuildHistoryLimit = &defaultSuccessfulBuildHistoryLimit
+	if i.Spec.SuccessBuildHistoryLimit == nil {
+		i.Spec.SuccessBuildHistoryLimit = &defaultSuccessfulBuildHistoryLimit
 	}
 
-	if s.Spec.CacheSize == nil && ctx.Value(HasDefaultStorageClass) != nil {
-		s.Spec.CacheSize = &defaultCacheSize
+	if i.Spec.CacheSize == nil && ctx.Value(HasDefaultStorageClass) != nil {
+		i.Spec.CacheSize = &defaultCacheSize
 	}
 }
 
-func (s *Image) Validate(ctx context.Context) *apis.FieldError {
-	return s.Spec.Validate(ctx).ViaField("spec")
+func (i *Image) Validate(ctx context.Context) *apis.FieldError {
+	return i.Spec.Validate(ctx).ViaField("spec")
 }
 
 func (is *ImageSpec) Validate(ctx context.Context) *apis.FieldError {
 	return is.validateTag(ctx).
 		Also(validateBuilder(is.Builder).ViaField("builder")).
 		Also(is.Source.Validate(ctx).ViaField("source")).
-		Also(is.Build.Validate(ctx).ViaField("build"))
+		Also(is.Build.Validate(ctx).ViaField("build")).
+		Also(is.validateCacheSize(ctx))
 }
 
-func (im *ImageSpec) validateTag(ctx context.Context) *apis.FieldError {
+func (is *ImageSpec) validateTag(ctx context.Context) *apis.FieldError {
 	if apis.IsInUpdate(ctx) {
 		original := apis.GetBaseline(ctx).(*Image)
-		return validate.ImmutableField(original.Spec.Tag, im.Tag, "tag")
+		return validate.ImmutableField(original.Spec.Tag, is.Tag, "tag")
 	}
 
-	return validate.Tag(im.Tag)
+	return validate.Tag(is.Tag)
+}
+
+func (is *ImageSpec) validateCacheSize(ctx context.Context) *apis.FieldError {
+	if apis.IsInUpdate(ctx) {
+		original := apis.GetBaseline(ctx).(*Image)
+		return validate.ImmutableField(
+			*original.Spec.CacheSize,
+			*is.CacheSize,
+			"cacheSize",
+			fmt.Sprintf("got: %v, want: %v", is.CacheSize, original.Spec.CacheSize),
```

Comment:
How is this different from the ImmutableField details?

Reply:
@matthewmcnew 

`k8s.io/apimachinery/pkg/api/resource.Quantity` only has pointer receiver implementation so since we need to use the value for comparison the `%v` format would be `{{6 9} {<nil>} 6G DecimalSI}`

We also could have compared `k8s.io/apimachinery/pkg/api/resource.Quantity.Value()` or `k8s.io/apimachinery/pkg/api/resource.Quantity.AsDec()` but then the error would read ex: `got: 6000000000, want: 5000000000` which I think is less readable.

I think this is an example of an interface that could plausibly show up again for other immutable fields. Also, it's optional (although a little ugly with `strings.Join(errDetails, "\n")`)